### PR TITLE
Adding GameController controls in Thunderscope

### PR DESCRIPTION
**Pull request by sauravbanna**
### Description

Added a new toolbar which allows us to send some GameController commands through the Thunderscope UI

image
image
image

### Testing Done

Thunderscope builds and runs as normal. The command is correctly sent (verified through printing). The robots do what they're supposed to

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/proto/manual_gc_command.proto [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+package SSLProto;
+
+import "proto/ssl_gc_state.proto";
+import "proto/ssl_gc_geometry.proto";
+
+message ManualGCCommand
+{
+    required Command manual_command             = 1;
+    optional Vector2 final_ball_placement_point = 3;
```
Comment (nimazareian) [inline src/proto/manual_gc_command.proto]: Any reason why we're not using `= 2` here?

File: src/software/thunderscope/gl/gl_widget.py [hunk 2]
```diff
@@ -57,25 +63,36 @@ def __init__(
...
+        self.toolbars_menu = QMenu()
         self.layers_menu_actions = {}
         self.toolbar = GLFieldToolbar(
+            parent=self.gl_view_widget,
             on_camera_view_change=self.set_camera_view,
             on_measure_mode=self.toggle_measure_mode,
             layers_menu=self.layers_menu,
+            toolbars_menu=self.toolbars_menu,
             sandbox_mode=sandbox_mode,
         )
```
Comment (nimazareian) [inline src/software/thunderscope/gl/gl_widget.py]: Now with two toolbars, we should consider renaming this to something like `self.simulation_control_toolbar = GLSimulationControlToolbar( ... )`

File: src/software/thunderscope/gl/widgets/gl_gamecontroller_toolbar.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        elif play == GamecontrollerPlays.INDIRECT:
+            command_type = Command.Type.INDIRECT
+        elif play == GamecontrollerPlays.KICKOFF:
+            command_type = Command.Type.KICKOFF
+        else:
+            command_type = Command.Type.PENALTY
+
+        self.__send_gc_command(
+            command_type, (SslTeam.YELLOW if is_blue else SslTeam.BLUE),
+        )
```
Comment (nimazareian) [inline src/software/thunderscope/gl/widgets/gl_gamecontroller_toolbar.py]: This color logic seems to be flipped

File: src/software/thunderscope/gl/widgets/toolbar_icons/gamecontroller/icon_loader.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        GLGamecontrollerToolbarIconLoader.BLUE_ICON = get_icon(
+            "software/thunderscope/gl/widgets/toolbar_icons/gamecontroller/blue.svg"
+        )
+
+    return GLGamecontrollerToolbarIconLoader.BLUE_ICON
+
+
+def get_open_window_icon(color: str) -> QtGui.QPixmap:
+    """
+    Loads the Blue icon pixmap as a GLGamecontrollerToolbarIconLoader attribute
```
Comment (nimazareian) [inline src/software/thunderscope/gl/widgets/toolbar_icons/gamecontroller/icon_loader.py]: minor nit: The comment here seems to be from a different function

File: src/software/thunderscope/thunderscope_config.py [hunk 5]
```diff
@@ -355,11 +351,6 @@ def configure_two_ai_gamecontroller_view(
                 ),
                 refresh_func_counter=yellow_refresh_func_frametime_counter,
             ),
-            TScopeWebTab(
-                name="Gamecontroller",
-                key=TabNames.GAMECONTROLLER,
-                url=GAME_CONTROLLER_URL,
-            ),
```
Comment (nimazareian) [inline src/software/thunderscope/thunderscope_config.py:362]: :fire:

**Review by nimazareian — COMMENTED** [review]
I just tried it out and overall it looks great! Big fan of the icons and the transparency of the toolbars! I had some initial thoughts after a quick pass of some of the core files:

1. Looking at the Referee widget, it looks like we're not continuously receiving referee packets, since the timestamp is not increasing unless I press one of the gamecontroller buttons
2. The "PLAYS" dropdown seem to be working on the reverse team. Selecting blue items actually sends yellow events, and vise-versa (verified with the GC gui)
3. I remember that a while ago we had a discussion about possibly having game controller running in a custom port during simulation or field testing (maybe triggered with a new cmd line arg to launch gamecontroller (e.g. `--launch_gc`) with `--run_blue`/`--run_yellow`) so we could use the gamecontroller on the same network as other teams without them receiving our game controller's packets (note that we must be using the assigned port by default when running during an actual game!). From an initial glance it looks like this might be skipped? If you think its out of scope for this PR, feel free to create an issue and we could tackle it in a separate PR, though this would be a requried feature prior to Robocup.
4. To add to the above point, we shouldn't have the GC toolbar when we're in an actual game (i.e. whenever we're not running our own custom GameController). On that note, we should not have pause button as well I think
https://github.com/UBC-Thunderbots/Software/blob/f0efeafbb0b4ac14bbe108c75384d37cab2493d0/src/software/thunderscope/gl/widgets/gl_field_toolbar.py#L131-L134

**Comment by sauravbanna** [comment]
> I just tried it out and overall it looks great! Big fan of the icons and the transparency of the toolbars! I had some initial thoughts after a quick pass of some of the core files:
> 
> 1. Looking at the Referee widget, it looks like we're not continuously receiving referee packets, since the timestamp is not increasing unless I press one of the gamecontroller buttons
> 2. The "PLAYS" dropdown seem to be working on the reverse team. Selecting blue items actually sends yellow events, and vise-versa (verified with the GC gui)
> 3. I remember that a while ago we had a discussion about possibly having game controller running in a custom port during simulation or field testing (maybe triggered with a new cmd line arg to launch gamecontroller (e.g. `--launch_gc`) with `--run_blue`/`--run_yellow`) so we could use the gamecontroller on the same network as other teams without them receiving our game controller's packets (note that we must be using the assigned port by default when running during an actual game!). From an initial glance it looks like this might be skipped? If you think its out of scope for this PR, feel free to create an issue and we could tackle it in a separate PR, though this would be a requried feature prior to Robocup.
> 4. To add to the above point, we shouldn't have the GC toolbar when we're in an actual game (i.e. whenever we're not running our own custom GameController). On that note, we should not have pause button as well I think
>    https://github.com/UBC-Thunderbots/Software/blob/f0efeafbb0b4ac14bbe108c75384d37cab2493d0/src/software/thunderscope/gl/widgets/gl_field_toolbar.py#L131-L134

1. I'll discuss on Discord about this but yea good catch
2. Fixed
3. I made a new issue for this: #3160, I'll get it done after this one
4. Yeah, that makes sense. I'll figure out how to do this, I'll discuss more on discord.

**Comment by nimazareian** [comment]
nit: Everything related to `QWebEngineView` should be removed (including `TScopeWebTab`) ! Including our pip dependency on it (including the `...requirement.txt` files).

File: src/software/thunderscope/gl/gl_widget.py (hunk 2 above)
Comment (nimazareian) [reply src/software/thunderscope/gl/gl_widget.py]: Ideally the class name is also updated

**Comment by nimazareian** [comment]
Just had one more request and an outstanding comment. Other than those it looks good!

**Review by nimazareian — APPROVED** [review]

File: src/software/thunderscope/gl/widgets/gl_gamecontroller_toolbar.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+        :param proto_unix_io the ProtoUnixIO object to send the manual gamecontroller commands to
+        :param friendly_color_yellow True if yellow is friendly team, False if not
+        """
+        super(GLGamecontrollerToolbar, self).__init__(parent=parent)
+
+        self.proto_unix_io = proto_unix_io
+        self.friendly_color_yellow = friendly_color_yellow
+
+        # the currently selected play (starts as None)
+        # self.current_selected_play = GamecontrollerPlays.NONE
```
Comment (williamckha) [inline src/software/thunderscope/gl/widgets/gl_gamecontroller_toolbar.py:58]: was this meant to be removed

File: src/software/thunderscope/gl/widgets/icon_loader.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
+from pyqtgraph.Qt import QtGui
+import os
+
+
+def get_icon(path: os.PathLike, color: str = None) -> QtGui.QPixmap:
```
Comment (williamckha) [inline src/software/thunderscope/gl/widgets/icon_loader.py:5]: Unrelated, but have we thought about using an icon library in tscope (e.g. https://github.com/spyder-ide/qtawesome)?

**Review by williamckha — APPROVED** [review]
LGTM!

**Pull request merged**